### PR TITLE
[Experimental] Implemented Hostname AntiAffinity

### PR DESCRIPTION
**Pull request by ellistarn**
**1. Issue, if available:**
https://github.com/aws/karpenter/issues/942

**2. Description of changes:**
Implements hostname antiaffinity by translating these rules into topology rules.

Requested feedback:
- Are the limitations going to work for users? Do we need a fully fledged implementation?
- Am I missing edge cases where my assumptions don't hold?
- Is anyone able to build this from HEAD and test it out?

**3. How was this change tested?**

**4. Does this change impact docs?**
- [ ] Yes, PR includes docs updates
- [ ] Yes, issue opened: *link to issue*
- [ ] No

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: pkg/controllers/selection/antiaffinity.go [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+	if len(term.Namespaces) != 0 {
+		errs = errs.Also(apis.ErrDisallowedFields("namespaces"))
+	}
+	if !AllowedAntiAffinityKeys.Has(term.TopologyKey) {
+		errs = errs.Also(apis.ErrInvalidKeyName(fmt.Sprintf("%s not in %v", term.TopologyKey, AllowedAntiAffinityKeys.UnsortedList()), "topologyKey"))
+	}
+	return errs
+}
+
+// Transform pod anti affinity into topology rules
```
Comment (akestner) [inline pkg/controllers/selection/antiaffinity.go:55]: Oh this is interesting. Is this how this works under the hood in K8s as well or a new approach that gets us there quicker in Karpenter?

Comment (ellistarn) [reply pkg/controllers/selection/antiaffinity.go:55]: This is absolutely not how it works under the hood. It's a bit of a hack that takes advantage of simplifying antiaffinity enough until it's translatable to topology.

**Review by felix-zhe-huang — COMMENTED** [review]
From https://kubernetes.io/docs/concepts/scheduling-eviction/assign-pod-node/ I saw this.
>"Note: Pod anti-affinity requires nodes to be consistently labelled, in other words every node in the cluster must have an appropriate label matching topologyKey. If some or all nodes are missing the specified topologyKey label, it can lead to unintended behavior."

So it seems that we also need to create a mechanism to label the existing nodes with the topologKey in order for this to work.
